syntax = "proto3";

package bookstore;

// Serviço de Autenticação
service AuthService {
  rpc Register(User) returns (AuthResponse);
  rpc Login(User) returns (AuthResponse);
}

// Serviço de Catálogo de Livros
service BookCatalogService {
  rpc GetBooks(Empty) returns (BookList);
  rpc GetBookDetails(BookRequest) returns (Book);
  rpc GetBookDetailsByID(BookRequestByID) returns (Book);
}

// Serviço de Gestão de Pedidos
service OrderService {
  rpc PlaceOrder(OrderRequest) returns (OrderResponse);
  rpc GetOrderDetails(OrderId) returns (Order);
  rpc GetOrderHistory(UserId) returns (OrderHistory);
}

message User {
  string username = 1;
  string password = 2;
}

message AuthResponse {
  string token = 1;
  string message = 2;
}

message Empty {}

message Book {
  int32 id = 1;
  string title = 2;
  string author = 3;
  int32 year = 4;
  int32 stock = 5;
  double price = 6;
}

message BookRequest {
  string title = 1;
}

message BookRequestByID {
  int32 id = 1;
}

message BookList {
  repeated Book books = 1;
}

message OrderItem {
  int32 book_id = 1;
  int32 quantity = 2;
  Book book = 3;
}

message OrderRequest {
  string username = 1;
  repeated OrderItem order_items = 2;
}

message OrderResponse {
  string order_id = 1;
  string message = 2;
}

message OrderId {
  string order_id = 1;
}

message Order {
  string order_id = 1;
  string username = 2;
  repeated OrderItem order_items = 3;
  string order_date = 4;
}

message UserId {
  string username = 1;
}

message OrderHistory {
  repeated Order orders = 1;
}
